hudi-client/hudi-java-client/src/test/java/org/apache/hudi/client/functional/TestHoodieJavaClientOnMergeOnReadStorage.java

```diff
@@ -0,0 +1,122 @@
+  public void setUpTestTable() {
+    testTable = HoodieTestTable.of(metaClient);
+  }
+
+  @Test
+  public void testReadingMORTableWithoutBaseFile() throws Exception {
+    HoodieWriteConfig config = getConfigBuilder(HoodieTestDataGenerator.TRIP_EXAMPLE_SCHEMA,
+        HoodieIndex.IndexType.INMEMORY).withAutoCommit(true)
+        .withCompactionConfig(HoodieCompactionConfig.newBuilder().withMaxNumDeltaCommitsBeforeCompaction(2).build())
+        .build();
+    HoodieJavaWriteClient client = getHoodieWriteClient(config);
+
+    // Do insert and updates thrice one after the other.
+    // Insert
+    String commitTime = HoodieActiveTimeline.createNewInstantTime();
+    insertBatch(config, client, commitTime, "000", 100, HoodieJavaWriteClient::insert,
+        false, false, 100, 100, 1, Option.empty());
+
+    // Update
+    String commitTimeBetweenPrevAndNew = commitTime;
+    commitTime = HoodieActiveTimeline.createNewInstantTime();
+    updateBatch(config, client, commitTime, commitTimeBetweenPrevAndNew,
+        Option.of(Arrays.asList(commitTimeBetweenPrevAndNew)), "000", 50, HoodieJavaWriteClient::upsert,
+        false, false, 50, 100, 2, config.populateMetaFields());
+
+    // Delete 5 records
+    String prevCommitTime = commitTime;
+    commitTime = HoodieActiveTimeline.createNewInstantTime();
+    deleteBatch(config, client, commitTime, prevCommitTime, "000", 25, false, false,
+        0, 100);
+
+    // Verify all the records.
+    metaClient.reloadActiveTimeline();
+    Map<String, GenericRecord> recordMap = GenericRecordValidationTestUtils.getRecordsMap(config, hadoopConf, dataGen);
+    assertEquals(75, recordMap.size());
+  }
+
+  @Test
+  public void testCompactionOnMORTable() throws Exception {
+    HoodieWriteConfig config = getConfigBuilder(HoodieTestDataGenerator.TRIP_EXAMPLE_SCHEMA,
+        HoodieIndex.IndexType.INMEMORY).withAutoCommit(true)
+        .withCompactionConfig(HoodieCompactionConfig.newBuilder().withMaxNumDeltaCommitsBeforeCompaction(2).build())
+        .build();
+    HoodieJavaWriteClient client = getHoodieWriteClient(config);
+
+    // Do insert and updates thrice one after the other.
+    // Insert
+    String commitTime = HoodieActiveTimeline.createNewInstantTime();
+    insertBatch(config, client, commitTime, "000", 100, HoodieJavaWriteClient::insert,
+        false, false, 100, 100, 1, Option.empty());
+
+    // Update
+    String commitTimeBetweenPrevAndNew = commitTime;
+    commitTime = HoodieActiveTimeline.createNewInstantTime();
+    updateBatch(config, client, commitTime, commitTimeBetweenPrevAndNew,
+        Option.of(Arrays.asList(commitTimeBetweenPrevAndNew)), "000", 50, HoodieJavaWriteClient::upsert,
+        false, false, 5, 100, 2, config.populateMetaFields());
+
+    // Schedule and execute compaction.
+    Option<String> timeStamp = client.scheduleCompaction(Option.empty());
+    assertTrue(timeStamp.isPresent());
+    client.compact(timeStamp.get());
+
+    // Verify all the records.
```

Comment:
can we add just 1 test to mimic async compaction. 
1 batch: inserts
2 upsert
schedule compaction. 
3rd upsert // this should add log files to new phantom file slice created by schedule compaction. 
try querying the table. 
now lets execute the compaction. writecliient.compact()
// ensure compaction completes. 
again trigger another batch of ingestion. 
all records should be intact.

Reply:
Do we have a similar test for the spark client? I'll make sure to give it the same name for future test improvements

Reply:
I added the test btw